cdap-ui/app/cdap/components/ContextMenu/index.tsx

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+*/
+
+import * as React from 'react';
+import Menu from '@material-ui/core/Menu';
+import MenuItem from '@material-ui/core/MenuItem';
+import withStyles from '@material-ui/core/styles/withStyles';
+import PropTypes from 'prop-types';
+export interface IContextMenuOption {
+  name: string;
+  label: (() => string) | string;
+  onClick: (event: React.SyntheticEvent) => void;
+  disabled?: boolean;
+}
+
+interface IContextMenuProps {
+  element?: HTMLElement;
+  selector?: string;
+  options?: IContextMenuOption[];
+  onOpen?: () => void; // to update disabled flags
+}
+
+const initialMousePosition = {
+  mouseX: null,
+  mouseY: null,
+};
+
+const StyledMenuItem = withStyles(() => ({
+  root: {
+    minHeight: 'auto',
+  },
+}))(MenuItem);
+
+const StyledDisabledMenuItem = withStyles((theme) => ({
+  root: {
+    minHeight: 'auto',
+    color: theme.palette.grey[500],
+    cursor: 'not-allowed',
+  },
+}))(MenuItem);
+
+export const ContextMenu = ({ selector, element, options, onOpen }: IContextMenuProps) => {
+  const [mousePosition, setMousePosition] = React.useState(initialMousePosition);
+
+  const toggleMenu = (e: PointerEvent) => {
+    setMousePosition({
+      mouseX: e.clientX - 2,
+      mouseY: e.clientY - 4,
+    });
+    e.preventDefault();
+    e.stopPropagation();
+    e.stopImmediatePropagation();
+    if (onOpen) {
+      onOpen();
+    }
+  };
+  const defaultEventHandler = (e) => e.preventDefault();
+
+  // state to capture children of context menu to disable right click on them.
+  const [children, setChildren] = React.useState(null);
+  // we don't use 'useRef' but a 'useCallback' is because 'ref.current' state is not
+  // tracked. So a useEffect(() => {...}, [ref.current]) won't get called back if
+  // the dom node changes.
+  const measuredRef = React.useCallback((node) => {
+    if (node !== null) {
+      setChildren([...Array.prototype.slice.call(node.children), node]);
+    }
+  }, []);
+
+  React.useEffect(
+    () => {
+      if (children) {
+        children.forEach((child) => {
+          child.removeEventListener('contextmenu', defaultEventHandler);
+          child.addEventListener('contextmenu', defaultEventHandler);
+        });
+      }
+      return () => {
```

Comment:
what do you think of not doing this cleanup? This gets called when the DOM is about to be unmounted, which means the children will be removed from DOM anyways. Also this might prevent a potential error if the children already got unmounted and we are trying to call removeEventListener from a non-existing element